README.md
# Network_2021
SOGANG_UNIV_NETWORK_2021 Project 1~3
(copy right - Professor So)

# Project 1
Make a program that sends a HTTP request to a specific URL, and receives and saves the response. 

# Project 2
Make a  program that simulates two Network layer protocol. (Link State routing algirhtm, Distance Vector routing algorithm)

# Project 3
Make a program that simulates the CRC(Cyclic Redundancy Check) in physical Layer. Both generates a codeword from dataword, and vice versa.


HW_2/changes.txt
0 1 -999
0 1 1
1 2 -999
1 4 -999
1 4 3
1 2 4


HW_2/linkstate_20161566.cc
#include<stdio.h>
#include<iostream>
#include<queue>
#include<string.h>


#define MAX_NODE 100
#define MAX_COST 100
#define DISCONNECTED -999 // = 1 bigger than MAXIMUM COST
#define MAX_MESSAGE 1000
using namespace std;

FILE *fp1, *fp2, *fp3, *fout;
int n;	// number of nodes

typedef struct _Node{
		pair<int, int> rout_table[MAX_NODE];	// Routing Table
		vector< pair<int,int> > neighbor;	//Neighbor Nodes
}Node;
Node node[MAX_NODE];

void Simulate(){
		fseek(fp2,0,SEEK_SET);
		char temp[MAX_MESSAGE];
		char message[MAX_MESSAGE];
		while(fgets(temp,MAX_MESSAGE,fp2)!=NULL){
				int start,end,curr,cost;
				int pos=3;
				if(strlen(temp)==1)
						continue;
				sscanf(temp,"%d %d",&start,&end);
				if(start>=10)
						pos++;
				if(end>=10)
						pos++;
				strcpy(message,temp+pos);

				queue<int> hops;
				curr=start;
				cost=node[start].rout_table[end].second;


				// Do simulations
				if(node[curr].rout_table[end].first==-1){
						fprintf(fout,"from %d to %d cost infinite hops ureachable message%s",start,end,message);
				}
				else{
						while(curr != end){
								hops.push(curr);
								curr = node[curr].rout_table[end].first;
						}
						// Print simulations

						fprintf(fout,"from %d to %d ",start,end);
						fprintf(fout,"cost %d hops ",cost);
						while(!hops.empty()){
								fprintf(fout,"%d ",hops.front());
								hops.pop();
						}
						fprintf(fout,"message%s",message);
				}
		}
		fprintf(fout,"\n");
}




// Tie-Breaking rule 2
// First, try the lower cost
// Second, if cost is same, try the smaller ID router
struct cmp{
		bool operator() (pair<int, int> &a, pair<int, int> &b){
				if(a.second==b.second){
						return a.first < b.first;
				}
				else{
						return a.second < b.second;
				}
		}
};

void Dijkstras(int router_num){
		int num;
		priority_queue <pair <int, int>, vector<pair<int, int> >,cmp >PQ; 
		
		// first: next node, second: cost of expanding that node
		
		// 1) Push First Router
		PQ.push(make_pair(router_num,0));

		// 2) Pop Next Router
		while(!PQ.empty()){
				// 2) Pop Next router,
				int parent[MAX_NODE];
				int next_node = PQ.top().first;	// Next Spannable node 
				int next_cost = PQ.top().second;	// The cost to that spannable node
				PQ.pop();

				// Check neighbors of next_node, and update them
				num=node[next_node].neighbor.size();
				for(int i=0;i<num;i++){
						int test_node = node[next_node].neighbor[i].first;
						int test_cost = node[next_node].neighbor[i].second;

						if(node[router_num].rout_table[test_node].second > next_cost + test_cost || node[router_num].rout_table[test_node].second==DISCONNECTED){	// If new Route
								// Update Cost
								node[router_num].rout_table[test_node].second = next_cost + test_cost;
								// Update 'Next'
								parent[test_node]=next_node;
								if(node[router_num].rout_table[next_node].first == router_num)
										node[router_num].rout_table[test_node].first = test_node;
								else
									node[router_num].rout_table[test_node].first = node[router_num].rout_table[next_node].first;
								// Push to queue
								PQ.push(make_pair(test_node,next_cost+test_cost));
						}
						else if(node[router_num].rout_table[test_node].second == next_cost + test_cost){	// Tie-breaking rule 3
								int new_parent=next_node;
								int old_parent=parent[test_node];

								if(old_parent > new_parent){	// Compare Parent ID, Change Root Parent
										parent[test_node]=next_node;
										if(node[router_num].rout_table[next_node].first == router_num)
												node[router_num].rout_table[test_node].first = test_node;
										else
											node[router_num].rout_table[test_node].first = node[router_num].rout_table[next_node].first;
								}
								PQ.push(make_pair(test_node,next_cost+test_cost));
						}
				}
		}
}

void Print_Routing_Tables(){
		for(int i=0;i<n;i++){
				for(int j=0;j<n;j++){
						if(node[i].rout_table[j].first==-1)
								continue;
						fprintf(fout,"%d %d %d\n",j,node[i].rout_table[j].first,node[i].rout_table[j].second);
				}
				fprintf(fout,"\n");
		}
}


void Update_Routing_Table(){
		// Update Each Routing Table using Dijkstra algorithm
		for(int i=0;i<n;i++){
				for(int j=0;j<n;j++){
						if(i==j)
								node[i].rout_table[j]=make_pair(i,0);
						else
								node[i].rout_table[j]=make_pair(-1,DISCONNECTED);
				}
		}
		for(int i=0;i<n;i++)
				Dijkstras(i);
}

int main(int argc, char*argv[]){
		int temp1,temp2,temp3;

		if(argc!=4){
				fprintf(stderr,"usage: linkstate topologyfile messagesfile changesfile\n");
				return 1;
		}
		fp1=fopen(argv[1],"r");
		fp2=fopen(argv[2],"r");
		fp3=fopen(argv[3],"r");
		fout=fopen("output_ls.txt","w");

		if(fp1==NULL || fp2==NULL || fp3==NULL){
				fprintf(stderr,"Error: open input file.\n");
				return 1;
		}

		// STEP 0

		// 0-1) Make Initial Routing Table
		fscanf(fp1,"%d",&n);
		while(fscanf(fp1,"%d %d %d",&temp1,&temp2,&temp3) != EOF){
				node[temp1].neighbor.push_back(make_pair(temp2,temp3));
				node[temp2].neighbor.push_back(make_pair(temp1,temp3));
		}

		// 0-2) Update Routing Table Using Dijkstra
		Update_Routing_Table();
		Print_Routing_Tables();

		// 0-3) Simulate with message.txt
		Simulate();

		// STEP 1

		int change_start, change_end, change_value;
		while(fscanf(fp3, "%d %d %d",&change_start,&change_end,&change_value)!=EOF){
				if(change_value == DISCONNECTED){
						// Delete the neighbors of change_start, change_end
						int num=node[change_start].neighbor.size();
						for(int i=0;i<num;i++){
								if(node[change_start].neighbor[i].first==change_end){
										node[change_start].neighbor.erase(node[change_start].neighbor.begin()+i);
										break;
								}
						}
						num=node[change_end].neighbor.size();
						for(int i=0;i<num;i++){
								if(node[change_end].neighbor[i].first==change_start){
										node[change_end].neighbor.erase(node[change_end].neighbor.begin()+i);
										break;
								}
						}

				}
				else{
						int num=node[change_start].neighbor.size();
						for(int i=0;i<num;i++){
								if(node[change_start].neighbor[i].first == change_end){
										node[change_start].neighbor.erase(node[change_start].neighbor.begin()+i);
										break;
								}
						}
						num=node[change_end].neighbor.size();
						for(int i=0;i<num;i++){
								if(node[change_end].neighbor[i].first == change_start){
										node[change_end].neighbor.erase(node[change_end].neighbor.begin()+i);
										break;
								}
						}
						node[change_start].neighbor.push_back(make_pair(change_end,change_value));
						node[change_end].neighbor.push_back(make_pair(change_start,change_value));
				}
				Update_Routing_Table();
				Print_Routing_Tables();
				Simulate();
		}

		printf("Complete. Output file written to output_ls.txt.\n");
		
		fclose(fp1);
		fclose(fp2);
		fclose(fp3);
		fclose(fout);
		return 0;
}


HW_1/http_client_20161566.cc
#include<stdio.h>
#include<stdlib.h>
#include<unistd.h>
#include<errno.h>
#include<string.h>
#include<netdb.h>
#include<sys/types.h>
#include<netinet/in.h>
#include<sys/socket.h>
#include<arpa/inet.h>

#define MAX_INPUT 100
#define MAX_DATA 1024
//purpose: parse input, return 0 if unvalide input
//input: user input(http://~~)
//output: number of parsed elements + parsed result

int parse_input(char *input, char **return_hostname, char **return_port,char **return_filepath){
		int return_value=1;
		int length=strlen(input);
		char *temp;
		char *hostname= *return_hostname;
		char *port = *return_port;
		char *filepath = *return_filepath;
		char http[8]="http://";
		bool is_there_port = true;
		bool is_there_filepath = true;

		//check if it starts with http
		temp = (char*)malloc(sizeof(char)*length);
		strcpy(temp,input);
		temp[7]='\0';	
		if(strcmp(http,temp)){
				free(temp);
				return 0;
		}		
		//start parsing
		strcpy(temp, input+7);

		//is there port number?
		if(!strchr(temp,':')){
				is_there_port=false;
				strcpy(port,"80");
		}
		
		//is there file path?
		if(!strchr(temp,'/')){
				is_there_filepath=false;
				strcpy(filepath,"");
		}

		if(temp[strlen(temp)-1]=='/'){
				is_there_filepath=false;
				temp[strlen(temp)-1]='\0';
				strcpy(filepath,"");
		}

		if(is_there_port && is_there_filepath){ //host + port +filepath
				hostname = strtok(temp,":");
				port = strtok(NULL,"/");
				filepath = strtok(NULL," ");
				return_value=3;
		}
		else if(is_there_port){ // host + port
				hostname=strtok(temp,":");
				port=strtok(NULL," ");
				return_value=2;
		}
		else if(is_there_filepath){ //host + filepath
				hostname = strtok(temp,"/");
				filepath = strtok(NULL," ");
				return_value=2;
		}
		else{ // host
				strcpy(hostname,temp);
				return_value=1;
		}	
		
		strcpy(*return_hostname,hostname);
		strcpy(*return_port,port);
		strcpy(*return_filepath,filepath);
		return return_value;
}



int main(int argc, char *argv[]){
		int sockfd;
		int numbytes;
		int rv;
		int buf_last=0;
		char buf[MAX_DATA];
		char *print_buffer;
		char *temp;
		char *temp2;
		int content_length=0;

		bool header_flag=false;
		bool status_code_flag=false;
		bool content_length_flag=false;
		
		char *hostname=(char*)malloc(sizeof(char)*MAX_INPUT);
		char *port=(char*)malloc(sizeof(char)*MAX_INPUT);
		char *filepath=(char*)malloc(sizeof(char)*MAX_INPUT);

		struct addrinfo hints;
		struct addrinfo *servinfo;
		FILE *fp;

		//check if input is right
		if(argc!=2 || !parse_input(argv[1],&hostname,&port,&filepath)){
				fprintf(stderr,"usage: http_client http://hostname[:port][/path/to/file]\n");
				exit(1);
		}
		//------------------------------------------- connecting!
		memset(&hints,0,sizeof(hints));
		hints.ai_family=AF_UNSPEC;
		hints.ai_socktype=SOCK_STREAM;

		//get the server information using host name(DNS)
		if ((rv = getaddrinfo(hostname, port, &hints, &servinfo)) != 0){ 
				fprintf(stderr, "getaddrinfo: %s\n", gai_strerror(rv)); 
				return 1;
		}
		//setup socket
		if((sockfd=socket(servinfo->ai_family,servinfo->ai_socktype, servinfo->ai_protocol))==-1){
				perror("client: socket");
				return 2;
		}
		//try TCP connection
		if(connect(sockfd, servinfo->ai_addr, servinfo->ai_addrlen) == -1) {
				close(sockfd);
				perror("connect");
				exit(1);
		}
		//----------------------------------------- connecting finish!


		freeaddrinfo(servinfo);

		//----------------------------------------- creating HTTP request
		//first line
		memset(buf,0,MAX_DATA);
		strcpy(buf+buf_last,"GET");
		buf_last += strlen("GET");
		buf[buf_last++]=' ';
		buf[buf_last++]='/';

		strcpy(buf+buf_last,filepath);
		buf_last += strlen(filepath);
		buf[buf_last++]=' ';

		strcpy(buf+buf_last,"HTTP/1.1");
		buf_last += strlen("HTTP/1.1");
		buf[buf_last++]='\r';
		buf[buf_last++]='\n';
		
		//second line
		strcpy(buf+buf_last,"Host:");
		buf_last += strlen("Host:");
		buf[buf_last++]=' ';

		strcpy(buf+buf_last,hostname);
		buf_last += strlen(hostname);
		buf[buf_last++]=':';

		strcpy(buf+buf_last,port);
		buf_last += strlen(port);
		buf[buf_last++]='\r';
		buf[buf_last++]='\n';
		
		buf[buf_last++]='\r';
		buf[buf_last++]='\n';
		
		//-------------------------- created HTTP request!
				
		if(send(sockfd,buf,strlen(buf),0)==-1){
				perror("send");
				close(sockfd);
				exit(1);
		}
		
		//--------------------------------- receive result
		fp=fopen("20161566.out","w");
		print_buffer=(char*)malloc(sizeof(char)*MAX_DATA);
		temp=(char*)malloc(sizeof(char)*MAX_DATA);
		temp2=(char*)malloc(sizeof(char)*MAX_DATA);
		memset(temp,0,MAX_DATA);
		memset(print_buffer,0,MAX_DATA);
		
		int saved_bytes=0;
		while((numbytes=recv(sockfd,buf,MAX_DATA-1,0)) >= 0){
				buf[numbytes]='\0';
				saved_bytes+=numbytes;
				//header not finished yet!
				if(!header_flag){
						//Does header end in this buffer?
						if(strstr(buf,"\r\n\r\n")){
							header_flag=true;
						}
						//This is the first buffer!
						if(!status_code_flag){
								status_code_flag=true;
								strcpy(print_buffer,buf);
								strtok(print_buffer,"\r\n");
						}
						//Is there Content-length?
						if(!content_length_flag){
								strcpy(temp2,buf);
								if(temp=strstr(temp2,"content-length")){
										content_length_flag=true;
										strtok(temp,": ");
										temp = strtok(NULL,"\r\n ");
										content_length=atoi(temp);
								}
								if(temp=strstr(temp2,"Content-Length")){
										content_length_flag=true;
										strtok(temp,": ");
										temp = strtok(NULL,"\r\n ");
										content_length=atoi(temp);
								}
						}
						if(header_flag){
							if(!content_length_flag)
									break;
							strcpy(temp2,buf);
							temp=strstr(temp2,"\r\n\r\n");
							saved_bytes+=(strlen(temp)-strlen(temp2)-4);
							fprintf(fp,"%s",temp+4);
						}
				}		
				else
					fprintf(fp,"%s",buf);

		
				memset(buf,0,MAX_DATA);
				if(saved_bytes==content_length)
						break;
		}
		if(numbytes==-1){
				perror("recv");
				close(sockfd);
				fclose(fp);
				free(print_buffer);
				free(hostname);
				free(port);
				free(filepath);
				exit(1);
		}
		printf("%s\n",print_buffer);
		//if there is no content_length
		if(!content_length_flag){
				printf("Content-Length not specified.\n");
		}
		else{
				printf("%d bytes written to 20161566.out\n",content_length);
		}
		
		//---------------------------------- received data
		free(print_buffer);
		free(hostname);
		free(port);
		free(filepath);
		
		fclose(fp);
		close(sockfd);
		return 0;
}




HW_4/crc_decoder_20161566.cpp
#include<cstdio>
#include<algorithm>
#include<iostream>
#include<stdlib.h>
#include<string.h>
#include<string>

using namespace std;

FILE *fp1;	// input file
FILE *fp2;	// output file
FILE *fp3;	// result file

int dataword_size; // dataword size
int generator_len; // generator length
string generator;
string input="";
string result="";

int total_word;
int error_word;

void Decoding(char target){
		unsigned char compare = 0x80;
		while(compare>0){
				if((target&compare) != 0)
						input.append("1");
				else
						input.append("0");
				compare >>= 1;
		}
}

void delete_padding(unsigned char num){
		input = input.substr(num);
}

void write_decodefile(){
		int index=0;
		while(index != result.length()){
				unsigned char temp =0;
				unsigned char compare=0x80;
				for(int i=0;i<8;i++){
						if(result[index+i]=='1')
								temp += compare;
						compare >>= 1;
				}
				index += 8;
				fprintf(fp2,"%c",temp);
		}

}


int error_detection(string target){
		string curr = target.substr(0,generator_len);
		int index=0;
		while(index <= target.length() - generator_len){
				if(curr.length() != generator_len)
						curr += target.at(index+generator_len-1);

				if(curr.at(0) == '0'){
						index++;
						curr = curr.substr(1);
						continue;
				}

				for(int i=0;i<generator_len;i++){
						if(curr.at(i) == generator.at(i)){
								curr[i]='0';
						}
						else
								curr[i]='1';
				}
				curr=curr.substr(1);
				index++;
		}
		for(int i=0;i<curr.length();i++){
				if(curr.at(i)=='1'){
						return 1;
				}
		}
		return 0;
}

void dataword_extraction(){
		int granularity = dataword_size + generator_len-1;
		int index=0;
		while(index != input.length()){
				string temp = input.substr(index,granularity);
				error_word += error_detection(temp);
				total_word++;
				result.append(temp.substr(0,dataword_size));
				index += granularity;
		}

}


int main(int argc, char* argv[]){
		char curr;
		bool flag = false;
		unsigned char padding;
		// ---------------------- Exception Handling ---------------------

		if(argc != 6){
				printf("usage: ./crc_decoder input_file output_file result_file generator dataword_size\n");
				return -1;
		}

		fp1 = fopen(argv[1],"r");
		fp2 = fopen(argv[2],"w");
		fp3 = fopen(argv[3],"w");

		if(fp1 == NULL){		
				printf("input file open error.\n");
				return -1;
		}

		if(fp2 == NULL){
				printf("output file open error.\n");
				return -1;
		}

		if(fp3 == NULL){
				printf("result file open error.\n");
				return -1;
		}

		dataword_size = atoi(argv[5]);

		if(dataword_size != 4 && dataword_size !=8){
				printf("datawrod size must be 4 or 8.\n");
				return -1;
		}

		generator_len = strlen(argv[4]);
		generator = argv[4];

		// --------------------- Decoding -------------
		while(EOF != fscanf(fp1,"%c",&curr)){
				if(!flag){
						flag=true;
						padding=curr;
						continue;
				}
				Decoding(curr);
		}
		// ---------------------- Delete Padding ----------------------
		delete_padding(padding);
		// ----------------------- Extracting CodeWord  -----------------
		dataword_extraction();
		fprintf(fp3,"%d %d\n",total_word,error_word);
		// ---------------------- Write Decode File ---------------
		write_decodefile();


		fclose(fp1);
		fclose(fp2);
		fclose(fp3);
		return 0;
}



HW_4/crc_encoder_20161566.cpp
#include<cstdio>
#include<algorithm>
#include<iostream>
#include<stdlib.h>
#include<string.h>
#include<string>

using namespace std;

FILE *fp1;	// input file
FILE *fp2;	// output file

int dataword_size; // dataword size
int generator_len; // generator length
string generator;
string result="";


// Generating Dataword from character 
// Char -> String
string dataword_generator(char target){
		string value="";
		unsigned char compare;
		if(dataword_size==4)
				compare=0x08;
		else
				compare=0x80;

		while(compare>0){
				if((target & compare) != 0)
						value.append("1");
				else 
						value.append("0");
				compare >>= 1;
		}	
		return value;
}

// Modulo-2 Division (making codeword)
void codeword_generator(char target){
		string dataword = dataword_generator(target);
		dataword.append(generator_len-1,'0');
		string curr = dataword.substr(0,generator_len);
		int index=0;
		
		while(index <= dataword.length() - generator_len){
				if(curr.length() != generator_len)
					curr += dataword.at(index+generator_len-1);

				if(curr.at(0) == '0'){
						index++;
						curr = curr.substr(1);
						continue;
				}
				
				for(int i=0;i<generator_len;i++){
						if(curr.at(i) == generator.at(i)){
								curr[i]='0';
						}
						else
								curr[i]='1';
				}
				
				curr=curr.substr(1);
				index++;
		}
		result.append(dataword.substr(0,dataword_size));
		result.append(curr);
}


int main(int argc, char* argv[]){
		char curr;

		// ------------------ Exception Handling ----------------------
		
		if(argc != 5){
				printf("usage: ./crc_encoder input_file output_file generator dataword_size\n");
				return -1;
		}

		fp1 = fopen(argv[1],"r");
		fp2 = fopen(argv[2],"w");

		if(fp1 == NULL){		
				printf("input file open error.\n");
				return -1;
		}

		if(fp2 == NULL){
				printf("output file open error.\n");
				return -1;
		}

		dataword_size = atoi(argv[4]);

		if(dataword_size != 4 && dataword_size !=8){
				printf("datawrod size must be 4 or 8.\n");
				return -1;
		}

		generator_len = strlen(argv[3]);
		generator = argv[3];

		// --------------------- CodeWord Generating -----------------

		while(EOF != fscanf(fp1,"%c",&curr)){
				if(dataword_size == 4){
						char first = curr & 0xF0;
						char second = curr & 0x0F;
						first >>=4;
						codeword_generator(first);
						codeword_generator(second);
				}
				else{
						codeword_generator(curr);
				}

		}
		// ----------------- Padding ------------------------
		int pad = (8 - result.length()%8)%8;
		string padding = "";
		padding.append(pad,'0');
		padding.append(result);
		result = padding;
	
		// ----------------- Change String to Bit -------------------
		int len = result.length()/8;
		unsigned char temp = pad & 0xFF;
		fprintf(fp2,"%c",temp);
		for(int i=0;i<len;i++){
				unsigned char target = 0;
				unsigned char compare = 0x80;
				for(int j=0;j<8;j++){
						if(result[8*i+j]=='1')
								target |= compare;
						compare >>= 1;
				}
				fprintf(fp2,"%c",target);
		}

		fclose(fp1);
		fclose(fp2);
		return 0;
}



HW_2/distvec_20161566.cc
#include<stdio.h>
#include<iostream>
#include<queue>
#include<vector>
#include<string.h>

using namespace std;

#define MAX_NODE 100
#define MAX_COST 100
#define DISCONNECTED -999
#define THRESHOLD 9901
#define MAX_MESSAGE 1000

int n;	//node number
FILE *fp1,*fp2,*fp3,*fout;

queue<int> update;
typedef struct _Node{
		pair<int, int> rout_table[MAX_NODE];	// Routing Table
		vector< pair<int,int> > neighbor;	// Neighbor Nodes
}Node;
Node node[MAX_NODE];	

void Simulate(){
		fseek(fp2,0,SEEK_SET);	// Set Read pointer to start of file
		char temp[MAX_MESSAGE];
		char message[MAX_MESSAGE];
		while(fgets(temp,MAX_MESSAGE,fp2)!=NULL){
				int start,end,curr,cost;
				int pos=3;
				if(strlen(temp)==1)
						continue;
				sscanf(temp,"%d %d",&start,&end);
				if(start>=10)
						pos++;
				if(end>=10)
						pos++;
				strcpy(message,temp+pos);
				
				queue<int> hops;
				curr=start;
				cost=node[start].rout_table[end].second;

				// Do simulations
				if(node[curr].rout_table[end].first==-1){
						fprintf(fout,"from %d to %d cost infinite hops ureachable message%s",start,end,message);
				}
				else{
						while(curr != end){
								hops.push(curr);
								curr = node[curr].rout_table[end].first;
						}
						// Print simulations

						fprintf(fout,"from %d to %d ",start,end);
						fprintf(fout,"cost %d hops ",cost);
						while(!hops.empty()){
								fprintf(fout, "%d ",hops.front());
								hops.pop();
						}
						fprintf(fout, "message%s",message);
				}
		}
		fprintf(fout,"\n");
}

void Print_Routing_Tables(){
		for(int i=0;i<n;i++){
				for(int j=0;j<n;j++){
						if(node[i].rout_table[j].first==-1)
								continue;
						fprintf(fout,"%d %d %d\n",j, node[i].rout_table[j].first, node[i].rout_table[j].second);
				}
				fprintf(fout,"\n");
		}
}

void Update_Routing_Table(){
		// Trade Routing Tables with adjacent node

		while(!update.empty()){
				bool modified=false;
				int i = update.front();	// Node that sends Routing Table
				int num = node[i].neighbor.size();
				update.pop();

				for(int j=0;j<num;j++){
						int next = node[i].neighbor[j].first;	// for each neighbor node
						int cost = node[i].neighbor[j].second;	// cost to next neighbor node

						for(int k=0;k<n;k++){
								int new_cost = cost;
								int prev_cost = node[i].rout_table[k].second;
								
								if(node[next].rout_table[k].first==-1){	// there is no route(i-->next-->k)
										if(node[i].rout_table[k].first==next){
												node[i].rout_table[k]=make_pair(-1,DISCONNECTED);
												modified=true;
										}
										continue;
								}
								new_cost += node[next].rout_table[k].second; // there is route(i-->next + next-->k)
								if(new_cost > THRESHOLD){
										node[i].rout_table[k]=make_pair(-1,DISCONNECTED);
										modified=true;
										continue;
								}
								//printf("curr node: %d, neighbor: %d, destination: %d, prev_cost: %d, next_cost: %d\n",i,next,k,prev_cost,new_cost);

								// Update Cases
								// 1. new route is more cheap
								// 2. there was no route in the first place
								// 3. If the Received Routing table is from the 'next'

								if(new_cost < prev_cost || prev_cost == DISCONNECTED){
										node[i].rout_table[k].first = next;
										node[i].rout_table[k].second = new_cost;
										modified=true;	

								}
								else if(new_cost == prev_cost){ //Tie-breaking Rule 1 apply
										if(node[i].rout_table[k].first > next){
												node[i].rout_table[k].first=next;
												modified=true;
										}
								}
								else{
										if(node[i].rout_table[k].first == next){ // If previous Route passes next, it should be updated no matter what!
												node[i].rout_table[k].first = next;
												node[i].rout_table[k].second = new_cost;
												modified=true;

										}
								}
						}
				}
				if(modified){
						for(int j=0;j<num;j++){
								update.push(node[i].neighbor[j].first);
						}
				}
		}
}

int main(int argc, char*argv[]){
		int temp1,temp2,temp3;

		if(argc!=4){
				fprintf(stderr,"usage: distvec topologyfile messagesfile changesfile\n");
				return 1;
		}
		fp1=fopen(argv[1],"r");
		fp2=fopen(argv[2],"r");
		fp3=fopen(argv[3],"r");
		fout=fopen("output_dv.txt","w");

		if(fp1==NULL || fp2==NULL || fp3==NULL){
				fprintf(stderr,"Error: open input file.\n");
				return 1;
		}

		// STEP 0

		// 0-1) Make Initial Routing Table
		fscanf(fp1,"%d",&n);
		for(int i=0;i<n;i++){
				for(int j=0;j<n;j++){
						if(i==j)	
								node[i].rout_table[j]=make_pair(i,0);
						else	
								node[i].rout_table[j]=make_pair(-1,DISCONNECTED);
				}
		}

		while(fscanf(fp1,"%d %d %d",&temp1,&temp2,&temp3) != EOF){
				node[temp1].rout_table[temp2]=make_pair(temp2,temp3);	// Add To initial Routing Table
				node[temp2].rout_table[temp1]=make_pair(temp1,temp3);

				node[temp1].neighbor.push_back(make_pair(temp2,temp3));	// Add To Neighbor
				node[temp2].neighbor.push_back(make_pair(temp1,temp3));
		}

		// 0-2) Update Routing Table
		for(int i=0;i<n;i++)
				update.push(i);
		Update_Routing_Table();
		Print_Routing_Tables();
		// 0-3) Simulate with message.txt
		Simulate();

		// STEP 1
		int change_start, change_end, change_value;
		while(fscanf(fp3,"%d %d %d", &change_start,&change_end,&change_value)!=EOF){
				// 1-1) Apply One line of change to network
				// Disconnecting connection

				int num=node[change_start].neighbor.size();
				for(int i=0;i<num;i++){
						if(node[change_start].neighbor[i].first==change_end){
								node[change_start].neighbor.erase(node[change_start].neighbor.begin()+i);
								break;
						}
				}
				num=node[change_end].neighbor.size();
				for(int i=0;i<num;i++){
						if(node[change_end].neighbor[i].first==change_start){
								node[change_end].neighbor.erase(node[change_end].neighbor.begin()+i);
								break;
						}
				}

				// Updating New Connection

				if(change_value != DISCONNECTED){		
						node[change_start].neighbor.push_back(make_pair(change_end,change_value));
						node[change_end].neighbor.push_back(make_pair(change_start,change_value));
				}
				else{
						for(int i=0;i<n;i++){
								if(node[change_start].rout_table[i].first==change_end)
										node[change_start].rout_table[i]=make_pair(-1,DISCONNECTED);
								if(node[change_end].rout_table[i].first==change_start)
										node[change_end].rout_table[i]=make_pair(-1,DISCONNECTED);
						}
				}
				num = node[change_start].neighbor.size();
				for(int i=0;i<num;i++)
						update.push(node[change_start].neighbor[i].first);
				num=node[change_end].neighbor.size();
				for(int i=0;i<num;i++)
						update.push(node[change_end].neighbor[i].first);

				// 1-2) Update Routing Table, Print Routing Table Simulate
				update.push(change_start);
				update.push(change_end);
				Update_Routing_Table();
				Print_Routing_Tables();
				Simulate();
		}


		printf("Complete. Output file written to output_dv.txt.\n");

		fclose(fp1);
		fclose(fp2);
		fclose(fp3);
		fclose(fout);

		return 0;
}
